internal/client/main.go

```diff
@@ -2,71 +2,92 @@ package client
-		"Content-Type": "application/json",
-	})
 	c.resty.SetTransport(&http.Transport{
 		DialContext: (&net.Dialer{
 			KeepAlive: 3 * time.Second,
 		}).DialContext,
 		TLSHandshakeTimeout: 5 * time.Second,
-		TLSClientConfig:     &tls.Config{InsecureSkipVerify: insecureSkipVerify},
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: insecureSkipVerify},
 	})
+	c.resty.SetHostURL(vaultBaseUrl)
 	return c
 }
 
 type Client struct {
 	resty *resty.Client
-}
+	logger *zap.SugaredLogger
 
-func (c *Client) SetToken(token string) *Client {
-	c.resty.SetHeader("X-Vault-Token", token)
-	return c
+	vaultBaseUrl string
+
+	token string
+
+	approleLoginPath string
+	approleLogoutPath string
+	approleRoleId string
+	approleSecretId string
+
+	// approleResponse is the successful response from Vault after logging in using ApproleRoleId/ApproleSecretId
+	approleResponse *successResponse
+
+	// approleTokenExpiration the future date when the token expires
+	approleTokenExpiration *time.Time
 }
 
-func (c *Client) SetHostUrl(url string) *Client {
-	c.resty.SetHostURL(url)
-	return c
+func (c *Client) request() (*resty.Request) {
+	return c.resty.R().
+		SetHeaders(map[string]string{
+			"Accept":        "application/json",
+			"Content-Type":  "application/json",
+			"X-Vault-Token": c.getToken(),
+		})
 }
 
 func (c *Client) Get(path string, result, error interface{}) (*resty.Response, error) {
-	return c.resty.R().SetResult(result).SetError(error).Get(path)
+	return c.request().SetResult(result).SetError(error).Get(path)
 }
 
 func (c *Client) List(path string, result, error interface{}) (*resty.Response, error) {
-	return c.resty.R().SetResult(result).SetError(error).Execute("LIST", path)
+	return c.request().SetResult(result).SetError(error).Execute("LIST", path)
 }
 
 func (c *Client) Put(path string, body, result, error interface{}) (*resty.Response, error) {
-	return c.resty.R().SetBody(map[string]interface{}{"data": body}).SetResult(result).SetError(error).Put(path)
+	return c.request().SetBody(map[string]interface{}{"data": body}).SetResult(result).SetError(error).Put(path)
 }
 
 func (c *Client) Post(path string, body, result, error interface{}) (*resty.Response, error) {
-	return c.resty.R().SetBody(map[string]interface{}{"data": body}).SetResult(result).SetError(error).Post(path)
+	return c.request().SetBody(map[string]interface{}{"data": body}).SetResult(result).SetError(error).Post(path)
 }
 
 func (c *Client) ApproleLogin(path string, body, result, error interface{}) (*resty.Response, error) {
-	return c.resty.R().SetBody(body).SetResult(result).SetError(error).Post(path)
+	return c.request().SetBody(body).SetResult(result).SetError(error).Post(path)
 }
 
 func (c *Client) ApproleLogout(path string, body, result, error interface{}) (*resty.Response, error) {
-	return c.resty.R().SetBody(body).SetResult(result).SetError(error).Post(path)
+	return c.request().SetBody(body).SetResult(result).SetError(error).Post(path)
```

Comment:
The reason I asked previously about the approle auth vs token auth login is you have `c.request()` here, which has a `.getToken()` call, which then does `c.login()`, which has `c.ApproleLogin()`, and we are now back to calling `c.request()` again.

To test your changes, I would recommend doing the following to make sure everything is passing before review (I will update the README.md at some point with this info):

- Make sure you have `certmagic-vault-storage` and `caddy-vault-storage` repos checked out alongside each other (e.g.: `~/projects/certmagic-vault-storage` + `~/projects/caddy-vault-storage`.
- In the file `caddy-vault-storage/go.mod, uncomment the `//replace` line.
- In one terminal, run `make localdev`, which will start up `vault` in dev mode and configure some stuff.
- In another terminal, run `make test` to make sure you are passing all tests.
- Before you are ready to push changes & review, make sure to comment out the `replace` line you previously uncommented above.

Reply:
@sundbry did you have a chance to check out the above note?